Mathlib/RingTheory/Invariant.lean

```diff
@@ -0,0 +1,107 @@
+/-
+Copyright (c) 2024 Thomas Browning. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Thomas Browning
+-/
+import Mathlib.Algebra.Polynomial.GroupRingAction
+import Mathlib.RingTheory.IntegralClosure.Algebra.Defs
+
+/-!
+# Invariant Extensions of Rings
+
+Given an extension of rings `B/A` and an action of `G` on `B`, we introduce a predicate
+`Algebra.IsInvariant A B G` which states that every fixed point of `B` lies in the image of `A`.
+
+## Main statements
+* `Algebra.IsInvariant.isIntegral`: If `G` is finite, then `B/A` is an integral extensions.
+
+TODO: Prove the existence of Frobenius elements in this general setting.
+
+-/
+
+namespace Algebra
+
+variable (A B G : Type*) [CommSemiring A] [Semiring B] [Algebra A B]
+  [Group G] [MulSemiringAction G B]
+
+/-- An action of a group `G` on an extension of rings `B/A` is invariant if every fixed  -/
+class IsInvariant : Prop where
+  isInvariant : ∀ b : B, (∀ g : G, g • b = b) → ∃ a : A, algebraMap A B a = b
```

Comment:
```suggestion
/-- An action of a group `G` on an extension of rings `B/A` is invariant if every fixed  -/
@[mk_iff]
class IsInvariant : Prop where
  isInvariant : ∀ b : B, (∀ g : G, g • b = b) → ∃ a : A, algebraMap A B a = b
```

`mk_iff` generates the `isInvariant_def` lemma below.
Also the docstring doesn't seem finished.

I wonder if stating it as `FixedPoints.subsemiring G B ≤ (algebraMap A B).srange` would be useful but you'll have to define `FixedPoints.subsemiring` first, so maybe not.

And also I'm surprised you don't need equality here.

Reply:
Equality can be achieved by adding a `SMulCommClass` assumption. I'll update the docstring.